crates/polars-core/src/chunked_array/ops/gather.rs

```diff
@@ -0,0 +1,167 @@
+use arrow::array::Array;
+use polars_error::{polars_bail, polars_ensure, PolarsResult};
+
+use crate::chunked_array::ops::{ChunkTake, ChunkTakeUnchecked};
+use crate::chunked_array::ChunkedArray;
+use crate::datatypes::{IdxCa, PolarsDataType, StaticArray};
+use crate::prelude::*;
+use crate::utils::index_to_chunked_index;
+
+impl<T: PolarsDataType, I: AsRef<[IdxSize]> + ?Sized> ChunkTake<I> for ChunkedArray<T>
+where
+    ChunkedArray<T>: ChunkTakeUnchecked<I>,
+{
+    /// Gather values from ChunkedArray by index.
+    fn take(&self, indices: &I) -> PolarsResult<Self> {
+        let len = self.len();
+        let all_valid = indices.as_ref().iter().all(|i| (*i as usize) < len);
+        polars_ensure!(all_valid, ComputeError: "invalid index in gather");
+
+        // SAFETY: we just checked the indices are valid.
+        Ok(unsafe { self.take_unchecked(indices) })
+    }
+}
+
+unsafe fn target_value_unchecked<'a, A: StaticArray>(
+    targets: &[&'a A],
+    idx: IdxSize,
+) -> A::ValueT<'a> {
+    let (chunk_idx, arr_idx) =
+        index_to_chunked_index(targets.iter().map(|a| a.len()), idx as usize);
+    let arr = targets.get_unchecked(chunk_idx);
+    arr.value_unchecked(arr_idx)
+}
+
+unsafe fn target_get_unchecked<'a, A: StaticArray>(
+    targets: &[&'a A],
+    idx: IdxSize,
+) -> Option<A::ValueT<'a>> {
+    let (chunk_idx, arr_idx) =
+        index_to_chunked_index(targets.iter().map(|a| a.len()), idx as usize);
+    let arr = targets.get_unchecked(chunk_idx);
+    arr.get_unchecked(arr_idx)
+}
+
+unsafe fn gather_idx_array_unchecked<A: StaticArray>(
+    dtype: DataType,
+    targets: &[&A],
+    has_nulls: bool,
```

Comment:
If we add `fn null_count` to `StaticArray` we could do the `has_null` branch per array. The null counts are cached, so the evaluation of the predicate is free.

Reply:
Right, I'll do that before merging this.

Reply:
@ritchie46 Actually, `StaticArray` already has `null_count` through `Array`. But even if only 1 target array has nulls, we must always use the null branch. The reason I passed `has_nulls` is because it's cheaper to test if any of the arrays has nulls on the CA.

Reply:
Yeap, I see. We go all over the chunks when we have multiple chunks. NVM :)